File: javascript/json-transform/src/functions/TransformerFunctionJoin.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { getAsString, isNullOrUndefined } from "../JsonHelpers";

class TransformerFunctionJoin extends TransformerFunction {
  constructor() {
    super({
      argsSets: [
        [
          { name: "delimiter", type: ArgType.String, defaultValue: "", aliases: ["$$delimiter"] },
          { name: "prefix", type: ArgType.String, defaultValue: "" },
          { name: "suffix", type: ArgType.String, defaultValue: "" },
          { name: "keep_nulls", type: ArgType.Boolean, defaultValue: false },
        ],
      ],
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null) return null;

    let delimiter = await context.getString("$$delimiter"); // backwards compat.
    if (delimiter == null) {
      delimiter = await context.getString("delimiter");
    }
    const prefix = (await context.getString("prefix")) ?? undefined;
    const suffix = (await context.getString("suffix")) ?? undefined;
    let stream = streamer.stream().map(getAsString);
    if (!(await context.getBoolean("keep_nulls"))) {
      stream = stream.filter(el => !isNullOrUndefined(el));
    }
    return stream.joinToString({
      prefix: prefix,
      postfix: suffix,
      separator: delimiter ?? "",
    });
  }
}

export default TransformerFunctionJoin;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionJoin.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;
import ai.bizone.jsontransform.functions.common.*;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class TransformerFunctionJoin extends TransformerFunction {
    public TransformerFunctionJoin() {
        super(FunctionDescription.of(
            Map.of(
            "delimiter", ArgumentType.of(ArgType.String).position(0).defaultValue("").aliases(List.of("$$delimiter")),
            "prefix", ArgumentType.of(ArgType.String).position(1).defaultValue(""),
            "suffix", ArgumentType.of(ArgType.String).position(2).defaultValue(""),
            "keep_nulls", ArgumentType.of(ArgType.Boolean).position(3).defaultValue(false)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var arr = context.getJsonElementStreamer(null);
        var delimiter = context.getString("$$delimiter"); // backwards compat.
        if (delimiter == null) {
            delimiter = context.getString("delimiter");
        }
        var prefix = Optional.ofNullable(context.getString("prefix")).orElse("");
        var suffix = Optional.ofNullable(context.getString("suffix")).orElse("");
        var stream = arr.stream().map(context::getAsString);
        if (!context.getBoolean("keep_nulls")) {
            stream = stream.filter(Objects::nonNull);
        }
        return stream.collect(Collectors.joining(delimiter, prefix, suffix));
    }
}
